common/fe-ste-models.go

```diff
@@ -1368,3 +1368,94 @@ func (p PreservePermissionsOption) IsTruthy() bool {
 		panic("unknown permissions option")
 	}
 }
+
+// CpkScopeInfo specifies the name of the encryption scope to use to encrypt the data provided in the request.
+// If not specified, encryption is performed with the default account encryption scope.
+// For more information, see Encryption at Rest for Azure Storage Services.
+type CpkScopeInfo struct {
+	EncryptionScope *string
+}
+
+func (csi CpkScopeInfo) Marshal() (string, error) {
+	result, err := json.Marshal(csi)
+	if err != nil {
+		return "", err
+	}
+	return string(result), nil
+}
+
+//func UnmarshalToCpkScopeInfo(cpkScopeInfoStr string) (CpkScopeInfo, error) {
+//	var result CpkScopeInfo
+//	if cpkScopeInfoStr != "" {
+//		err := json.Unmarshal([]byte(cpkScopeInfoStr), &result)
+//		if err != nil {
+//			return result, err
+//		}
+//	}
+//
+//	return result, nil
+//}
+
+type CpkInfo struct {
+	// The algorithm used to produce the encryption key hash.
+	// Currently, the only accepted value is "AES256".
+	// Must be provided if the x-ms-encryption-key header is provided.
+	EncryptionAlgorithm *string
+
+	// Optional. Specifies the encryption key to use to encrypt the data provided in the request.
+	// If not specified, encryption is performed with the root account encryption key.
+	EncryptionKey *string
+
+	// The SHA-256 hash of the provided encryption key.
+	// Must be provided if the x-ms-encryption-key header is provided.
+	EncryptionKeySha256 *string
+}
+
+func (csi CpkInfo) Marshal() (string, error) {
+	result, err := json.Marshal(csi)
+	if err != nil {
+		return "", err
+	}
+	return string(result), nil
+}
+
+//func UnmarshalToCpkInfo(cpkInfoStr string) (CpkInfo, error) {
+//	var result CpkInfo
+//	if cpkInfoStr != "" {
+//		err := json.Unmarshal([]byte(cpkInfoStr), &result)
+//		if err != nil {
+//			return result, err
+//		}
+//	}
+//
+//	return result, nil
+//}
+
+func ToClientProvidedKeyOptions(cpkInfo CpkInfo, cpkScopeInfo CpkScopeInfo) azblob.ClientProvidedKeyOptions {
+	return azblob.ClientProvidedKeyOptions{
```

Comment:
>azblob.ClientProvidedKeyOptions [](start = 8, length = 31)

minor style issue: wouldn't it be easier to use azblob.ClientProvidedKeyOptions directly to hold the data in the first place?

Reply:
This is inline with track2 changes which we're going to bring in and inline with other SDKs as well. There CpkScopeInfo and CpkInfo are treated as separate entities. To avoid that change, I've made sure to use two objects instead of one,